### PR TITLE
feat: Add support for EMR on EKS Fargate

### LINKED ISSUE #989: Creating multiple namespaces in emr-on-eks not working as expected
## Description

Hello,

I am using the examples/emr-on-eks blueprint. When trying to create multiple namespaces / virtual clusters, I run into an error. 

`╵
╷
│ Error: failed creating IAM Role (emr-on-eks-emr-on-eks-job-role): EntityAlreadyExists: Role with name emr-on-eks-emr-on-eks-job-role already exists.
│ 	status code: 409, request id: 4bb98f72-04cf-40cd-90ee-8405b49d1dc6
│ 
│   with module.eks_blueprints.module.emr_on_eks["jian"].aws_iam_role.emr_on_eks_execution,
│   on .terraform/modules/eks_blueprints/modules/emr-on-eks/main.tf line 88, in resource "aws_iam_role" "emr_on_eks_execution":
│   88: resource "aws_iam_role" "emr_on_eks_execution" {
│ 
╵
╷
│ Error: failed creating IAM Role (emr-on-eks-emr-on-eks-job-role): EntityAlreadyExists: Role with name emr-on-eks-emr-on-eks-job-role already exists.
│ 	status code: 409, request id: 6257de23-9b03-4f33-a929-b1b3250d953e
│ 
│   with module.eks_blueprints.module.emr_on_eks["jenn"].aws_iam_role.emr_on_eks_execution,
│   on .terraform/modules/eks_blueprints/modules/emr-on-eks/main.tf line 88, in resource "aws_iam_role" "emr_on_eks_execution":
│   88: resource "aws_iam_role" "emr_on_eks_execution" {
│ 
`

## Reproduction Code [Required]
This is what I mean by creating multiple namespaces and virtual clusters in main.tf

```
emr_on_eks_teams = {
    jenn= {
      namespace               = "jenn"
    },
    jian = {
      namespace               = "jian"
    }

```
```

resource "aws_emrcontainers_virtual_cluster" "jian" {
  name = format("%s-%s", module.eks_blueprints.eks_cluster_id, "jian")

  container_provider {
    id   = module.eks_blueprints.eks_cluster_id
    type = "EKS"

    info {
      eks_info {
        namespace = "jian"
      }
    }
  }
}


resource "aws_emrcontainers_virtual_cluster" "jenn" {
  name = format("%s-%s", module.eks_blueprints.eks_cluster_id, "jenn")

  container_provider {
    id   = module.eks_blueprints.eks_cluster_id
    type = "EKS"

    info {
      eks_info {
        namespace = "jenn"
      }
    }
  }
}
```

As a workaround, i append a random string to the emr-on-eks-emr-on-eks-job-role, so that multiple iam roles are created i.e. ... 

`resource "aws_iam_role" "emr_on_eks_execution" {
  name                  = format("%s-%s-%s", var.eks_cluster_id, local.emr_on_eks_team["job_execution_role"], **random_string.random.result)**
  assume_role_policy    = data.aws_iam_policy_document.emr_assume_role.json
  force_detach_policies = true
  path                  = var.iam_role_path
  permissions_boundary  = var.iam_role_permissions_boundary
  tags                  = var.tags
}`

### Issue
if i change the source in my code to use github path, then i cannot use the workaround I have stated (as it requires me to modify the modules/emr-on-eks/main.tf file. Possible you could include this workaround in the code / what is the best way to create multiple namespaces for this blueprint?

E.g. github path.
module "eks_blueprints" {
  source = "github.com/aws-ia/terraform-aws-eks-blueprints?ref=v4.10.0"

module "eks_blueprints_kubernetes_addons" {
  source = "github.com/aws-ia/terraform-aws-eks-blueprints//modules/kubernetes-addons?ref=v4.10.0"

### LINKED ISSUE #1027: EMR on EKS Fargate
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

#### What is the outcome that you are trying to reach?

[EMR on EKS](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples/analytics/emr-on-eks) but utilizing Fargate compute 

#### Describe the solution you would like

[EMR on EKS](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples/analytics/emr-on-eks) but utilizing Fargate compute

### PR DESCRIPTION
**Pull request by bryantbiggs**
### What does this PR do?

- Add new addon `emr-on-eks` to create EMR virtual clusters on EKS
- Add new `emr-on-eks-fargate` example/blueprint which demonstrates the new `emr-on-eks` addon plus provides a blueprint for running this workload on EKS Fargate
- Fix dependency ordering by propagating data source that waits for upstream dependency to resolve after a period of time; coordination between data plane and addon deployment
- Fix dependency ordering for CoreDNS cluster proportional autoscaler to ensure that CoreDNS is deployed first before deploying CPA

image

### Motivation

- Resolves #1027
- Resolves #989

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [x] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [ ] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

### CODE REVIEW

File: modules/kubernetes-addons/aws-coredns/main.tf [hunk 1]
```diff
@@ -193,6 +193,8 @@ module "cluster_proportional_autoscaler" {
         tolerations:
           - key: "CriticalAddonsOnly"
             operator: "Exists"
+        blueprints:
+          connection: ${try(null_resource.remove_default_coredns_deployment[0].id, "none")}
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/aws-coredns/main.tf:197]: just a hint to Terraform to better align the ordering of operations

File: modules/kubernetes-addons/locals.tf [hunk 2]
```diff
@@ -55,7 +55,7 @@ locals {
     aws_eks_cluster_endpoint       = local.eks_cluster_endpoint
     aws_partition_id               = data.aws_partition.current.partition
     aws_region_name                = data.aws_region.current.name
-    eks_cluster_id                 = var.eks_cluster_id
+    eks_cluster_id                 = data.aws_eks_cluster.eks_cluster.id
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/locals.tf:58]: this data source is tied to the `data_plane_wait_arn` which waits for the passed value to resolve before proceeding which helps to ensure the data plane is ready before the addons are provisioned

File: examples/analytics/emr-on-eks-fargate/README.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+Error: waiting for EMR Containers Virtual Cluster (xwbc22787q6g1wscfawttzzgb) delete: unexpected state 'ARRESTED', wanted target ''. last error: %!s(<nil>)
+```
+
+You can clean up any of the clusters in the `ARRESTED` state with the following:
+
+```sh
+aws emr-containers list-virtual-clusters --region us-west-2 --states ARRESTED \
+--query 'virtualClusters[0].id' --output text | xargs -I{} aws emr-containers delete-virtual-cluster \
+--region us-west-2 --id {}
+```
```
Comment (askulkarni2) [inline examples/analytics/emr-on-eks-fargate/README.md:94]: 👀

**Review by askulkarni2 — APPROVED** [review]
Awesome PR!

**Pull request merged**